File: Data Structures & Algorithms/maximum-subarray/submission-0.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        dp =[0]*len(nums) # initializing dp array

        dp[0]=nums[0] # base case

        for i in range(1,len(nums)):
            dp[i]=max(dp[i-1]+nums[i],nums[i])
        
        return max(dp)

        

File: Data Structures & Algorithms/house-robber/submission-0.java
class Solution {
    public int rob(int[] nums) {
        int dp[]=new int[nums.length];
        Arrays.fill(dp,-1);
        return rob(nums.length-1,nums,dp);
    }

    public int rob(int i, int[] nums, int[] dp){

        if(i==0){
            return nums[0];
        }

        if(i==1){
            return Math.max(nums[0], nums[1]);
        }

        if(dp[i]!=-1){
            return dp[i];
        }


        dp[i]=Math.max(nums[i] + rob(i-2,nums,dp), rob(i-1,nums,dp));
        return dp[i];

    }

}
